File: tools/contract-critic.py
#!/usr/bin/env python3
"""Check a goal's contract against the capabilities its parts actually import — before a run.

    python3 tools/contract-critic.py .comp/goals/treasury-ledger.toml

The cheap, deterministic half of what a branch did by hand twice in this experiment: it read a
contract, found a claim the components did not support, and filed a CONTRACT-REQUEST — after a
generation had already been spent. This catches the two shapes that cost real runs, with no
model call:

  * a context file that does not resolve (a dangling path is shown to no one and silently drops
    a capability from what the part can see);
  * a capability a part's WORLD imports whose call signature the contract never quotes, so the
    part must guess it — the class that produced app 6's uncovered `idempotency`.

It does NOT read component source to refute a contract's CLAIMS about behaviour — that is the
model-based critic, a larger and non-deterministic thing. This is the part that pays for itself
on every goal for free, and it is worth running before `goal-rehearse.sh`, which is slower.
"""
import re
import sys
import tomllib
from pathlib import Path

# package namespace -> the binding alias a contract quotes it under (e.g. money:amount -> money::)
ALIAS = {
    "records": "records", "money": "money", "ledger": "ledger", "idempotency": "idem",
    "ratelimit": "rl", "pii": "pii", "ai": "ai", "fsm": "fsm", "cache": "cache",
    "quota": "meter", "search": "search", "otp": "totp", "session": "sessions",
    "event": "bus", "policy": "policy", "outbox": "outbox", "notify": "notify",
    "audit": "audit", "auth": "authz", "lock": "mutex",
}


def check(goal_path: str) -> int:
    goal = tomllib.load(open(goal_path, "rb"))
    if "part" not in goal:
        print(f"{goal_path}: single-part goal, nothing to cross-check")
        return 0
    contract = Path(goal["contract"]).read_text()
    problems = []

    # 1. Every context path must resolve.
    for part in goal["part"]:
        for c in part.get("context", []):
            if not Path(c).exists():
                problems.append(f"{part['name']}: context path does not exist: {c}")

    # 2. Every capability a part's world imports should have its signature quoted in the contract.
    #    The world is the app's own .wit in the part's context.
    for part in goal["part"]:
        wits = [c for c in part.get("context", []) if c.endswith(".wit") and "-domain/" in c]
        for w in wits:
            if not Path(w).exists():
                continue
            for m in re.finditer(r"import ([a-z0-9-]+):", Path(w).read_text()):
                ns = m.group(1)
                if ns in ("wasi", "comp"):
                    continue
                alias = ALIAS.get(ns, ns)
                covered = bool(re.search(rf"\b{re.escape(alias)}::", contract)) or f"{ns}:" in contract
                if not covered:
                    problems.append(
                        f"{part['name']}: world imports `{ns}:` but the contract never quotes "
                        f"`{alias}::` — the part must guess the signature"
                    )

    problems = sorted(set(problems))
    title = goal.get("title", goal_path)
    if not problems:
        print(f"OK  {title}")
        return 0
    print(f"FAIL {title}")
    for p in problems:
        print(f"  · {p}")
    return 1


if __name__ == "__main__":
    if len(sys.argv) < 2:
        sys.exit("usage: contract-critic.py <goal.toml> [<goal.toml> ...]")
    rc = 0
    for g in sys.argv[1:]:
        rc |= check(g)
    sys.exit(rc)

File: xtask/src/contract_critic.rs
//! `cargo xtask contract-critic` — was tools/contract-critic.py.
//!
//! The cheap, deterministic half of what a branch used to do by hand: read a
//! goal's contract, find a claim the components do not support, before a
//! generation was spent on it. Two shapes, no model call:
//!   - a context file that does not resolve;
//!   - a capability a part's world imports whose signature the contract never
//!     quotes, so the part has to guess it.
//!
//! Ported from Python because it's a dev-tooling script with no component tie
//! — this repo's own convention keeps those in Rust alongside xtask.

use anyhow::{Context, Result};
use std::collections::BTreeSet;
use std::fs;
use std::path::{Path, PathBuf};

/// package namespace -> the binding alias a contract quotes it under
/// (e.g. `money:amount` is quoted as `money::`, `idempotency:*` as `idem::`).
fn contract_alias(ns: &str) -> &str {
    match ns {
        "idempotency" => "idem",
        "ratelimit" => "rl",
        "quota" => "meter",
        "otp" => "totp",
        "session" => "sessions",
        "event" => "bus",
        "auth" => "authz",
        "lock" => "mutex",
        other => other,
    }
}

/// The namespaces a `.wit` file's world imports (`import wasi:http/...;` -> `wasi`).
fn wit_imports(text: &str) -> Vec<String> {
    text.lines()
        .filter_map(|l| l.trim().strip_prefix("import "))
        .filter_map(|rest| rest.split(':').next())
        .map(str::to_string)
        .collect()
}

fn contract_critic_one(goal_path: &Path) -> Result<bool> {
    let goal: toml::Value = toml::from_str(
        &fs::read_to_string(goal_path)
            .with_context(|| format!("reading {}", goal_path.display()))?,
    )?;
    let title = goal
        .get("title")
        .and_then(|t| t.as_str())
        .map(str::to_string)
        .unwrap_or_else(|| goal_path.display().to_string());

    let Some(parts) = goal.get("part").and_then(|p| p.as_array()) else {
        println!("{}: single-part goal, nothing to cross-check", goal_path.display());
        return Ok(true);
    };
    let contract_path = goal
        .get("contract")
        .and_then(|c| c.as_str())
        .with_context(|| format!("{}: no `contract` key", goal_path.display()))?;
    let contract = fs::read_to_string(contract_path)
        .with_context(|| format!("reading contract {contract_path}"))?;

    let mut problems: BTreeSet<String> = BTreeSet::new();
    for part in parts {
        let name = part.get("name").and_then(|n| n.as_str()).unwrap_or("?");
        let context: Vec<&str> = part
            .get("context")
            .and_then(|c| c.as_array())
            .map(|a| a.iter().filter_map(|v| v.as_str()).collect())
            .unwrap_or_default();

        // 1. Every context path must resolve.
        for c in &context {
            if !Path::new(c).exists() {
                problems.insert(format!("{name}: context path does not exist: {c}"));
            }
        }

        // 2. Every capability a part's world imports should have its signature
        //    quoted in the contract — the world is the app's own .wit, among
        //    the part's context.
        for w in context.iter().filter(|c| c.ends_with(".wit") && c.contains("-domain/")) {
            let Ok(text) = fs::read_to_string(w) else { continue };
            for ns in wit_imports(&text) {
                if ns == "wasi" || ns == "comp" {
                    continue;
                }
                let alias = contract_alias(&ns);
                let covered = contract.contains(&format!("{alias}::"))
                    || contract.contains(&format!("{ns}:"));
                if !covered {
                    problems.insert(format!(
                        "{name}: world imports `{ns}:` but the contract never quotes \
                         `{alias}::` — the part must guess the signature"
                    ));
                }
            }
        }
    }

    if problems.is_empty() {
        println!("OK  {title}");
        Ok(true)
    } else {
        println!("FAIL {title}");
        for p in &problems {
            println!("  · {p}");
        }
        Ok(false)
    }
}

pub(crate) fn contract_critic(goals: &[PathBuf]) -> Result<()> {
    if goals.is_empty() {
        anyhow::bail!("usage: cargo xtask contract-critic <goal.toml> [<goal.toml> ...]");
    }
    let mut all_ok = true;
    for g in goals {
        if !contract_critic_one(g)? {
            all_ok = false;
        }
    }
    if all_ok {
        Ok(())
    } else {
        anyhow::bail!("one or more goals have contract problems");
    }
}
